src/libraries/System.IO.Compression/src/System/IO/Compression/DeflateZLib/Deflater.cs

```diff
@@ -24,10 +24,50 @@ internal sealed class Deflater : IDisposable
         // Note, DeflateStream or the deflater do not try to be thread safe.
         // The lock is just used to make writing to unmanaged structures atomic to make sure
         // that they do not get inconsistent fields that may lead to an unmanaged memory violation.
-        // To prevent *managed* buffer corruption or other weird behaviour users need to synchronise
+        // To prevent *managed* buffer corruption or other weird behavior users need to synchronize
         // on the stream explicitly.
         private object SyncLock => this;
 
+        internal Deflater(ZLibCompressionOptions options, int windowBits)
+        {
+            Debug.Assert(windowBits >= minWindowBits && windowBits <= maxWindowBits);
+
+            int memLevel = options.CompressionLevel == 0 ? ZLibNative.Deflate_NoCompressionMemLevel : ZLibNative.Deflate_DefaultMemLevel;
+            ZErrorCode errC;
+            try
+            {
+                errC = ZLibNative.CreateZLibStreamForDeflate(out _zlibStream, (ZLibNative.CompressionLevel)options.CompressionLevel, windowBits, memLevel, (ZLibNative.CompressionStrategy)options.CompressionStrategy);
+            }
+            catch (Exception cause)
+            {
+                throw new ZLibException(SR.ZLibErrorDLLLoadError, cause);
+            }
+
+            CheckErrorCode(errC);
+        }
+
+        private void CheckErrorCode(ZErrorCode errC)
+        {
+            switch (errC)
+            {
+                case ZErrorCode.Ok:
+                    return;
+
+                case ZErrorCode.MemError:
+                    throw new ZLibException(SR.ZLibErrorNotEnoughMemory, "deflateInit2_", (int)errC, _zlibStream.GetErrorMessage());
+
+                case ZErrorCode.VersionError:
+                    throw new ZLibException(SR.ZLibErrorVersionMismatch, "deflateInit2_", (int)errC, _zlibStream.GetErrorMessage());
+
+                case ZErrorCode.StreamError:
+                    throw new ZLibException(SR.ZLibErrorIncorrectInitParameters, "deflateInit2_", (int)errC, _zlibStream.GetErrorMessage());
+
+                default:
+                    throw new ZLibException(SR.ZLibErrorUnexpected, "deflateInit2_", (int)errC, _zlibStream.GetErrorMessage());
+            }
+        }
+
+
         internal Deflater(CompressionLevel compressionLevel, int windowBits)
```

Comment:
We shouldn't need to duplicate this ctor. Could we instead do something like this:
```C#
internal Deflator(int compressionLevel, ZLibCompressionStrategy compressionStrategy, int windowBits)
{
    ... // logic from the above ctor but using these direct parameters rather than the corresponding values from options
}

internal static int GetZLibCompressionLevel(CompressionLevel level) { ... }
internal static int GetMemLevel(CompressionLevel level) { ... }
```
and then whatever is currently calling the `Deflater(CompressionLevel compressionLevel, int windowBits)` can just call the single ctor with the results of those two helpers?